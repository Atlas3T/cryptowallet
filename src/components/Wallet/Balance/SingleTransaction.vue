<template>
  <q-timeline-entry
    :title="paymentDirection"
    side="left"
  >
    <q-collapsible
      :label="date"
      :sublabel="amount"
      :class="{ 'positive-amount': data.amount > 0, 'negative-amount': data.amount < 0 }"
    >
      <div>
        <p>
          {{ $t('received') }}:<br>
          <span class="received">{{ data.received }}</span>
        </p>
        <p>
          TX Hash:<br>
          <span class="tx-hash">{{ data.tx_hash }}</span>
        </p>
        <p>
          {{ $t('status') }}:<br>
          <span
            :class="{ 'unconfirmed-tx': data.status === 1, 'confirmed-tx': data.status === 2 }"
            class="status"
          >
            {{ $t(status) }}
          </span>
        </p>
        <p>
          TX {{ $t('fee') }}:<br>
          <span class="tx-fee">{{ data.tx_fee }}</span>
        </p>
      </div>
    </q-collapsible>
  </q-timeline-entry>
</template>

<script>
import { amount, date } from '@/helpers';

export default {
  name: 'SingleTransaction',
  props: {
    data: {
      type: Object,
      required: true,
    },
    currency: {
      type: String,
      default() {
        return '';
      },
    },
  },
  computed: {
    /**
     * Prepends "To: " or "From: " into transaction title
     * @return {String}
     */
    paymentDirection() {
      if (this.data.to) return `${this.$t('to')}: ${this.data.to}`;
      return `${this.$t('from')}: ${this.data.from}`;
    },
    /**
     * Returns formatted date of the transaction
     * @return {String}
     */
    date() {
      return date(this.data.ts, 'DD MMMM', this);
    },
    /**
     * Converts amount value (Number) into a String with optional currency symbol
     * @return {String}
     */
    amount() {
      return amount(this.data.amount, this.currency);
    },
    /**
     * Returns the transaction status in a human friendly format
     * @return {String}
     */
    status() {
      if (this.data.status === 1) return 'unconfirmed';
      if (this.data.status === 2) return 'confirmed';
      return 'pending';
    },
  },
};
</script>

<style>
  .confirmed-tx {
    color: green;
  }

  .unconfirmed-tx {
    color: red;
  }

  .positive-amount .q-item-sublabel {
    color: green;
  }

  .negative-amount .q-item-sublabel {
    color: red;
  }
</style>
