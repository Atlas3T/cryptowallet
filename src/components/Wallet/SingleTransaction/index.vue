<template>
  <q-timeline-entry
    :title="date"
    side="left"
  >
    <q-expansion-item
      :label="amount"
      :class="{ 'positive-amount': !data.sent, 'negative-amount': data.sent }"
    >
      <q-card>
        <q-card-section>
          <div class="single-transaction-content">
            <p>
              TX Hash:
              <span class="tx-hash break">{{ data.hash }}</span>
            </p>
            <p>
              {{ $t('status') }}:
              <span
                :class="{
                  'unconfirmed-tx': data.confirmations < minConfirmations,
                  'confirmed-tx': data.confirmations >= minConfirmations,
                }"
                class="status"
              >
                {{ status }}
              </span>
            </p>
            <p>
              {{ $t('confirmations') }}:
              <span class="confirmations">{{ data.confirmations }}</span>
            </p>
            <p>
              TX {{ $t('fee') }}:
              <span class="tx-fee">{{ feeFormated }}</span>
            </p>
            <p>
              {{ $t('blockHeight') }}:
              <span class="tx-height">{{ data.blockHeight }}</span>
            </p>
          </div>
        </q-card-section>
      </q-card>
    </q-expansion-item>

  </q-timeline-entry>
</template>

<script>
import { mapState } from 'vuex';
import { dateTranslater, AmountFormatter } from '@/helpers';
import Coin from '@/store/wallet/entities/coin';

export default {
  name: 'SingleTransaction',

  props: {
    data: {
      type: Object,
      required: true,
    },
  },

  computed: {
    ...mapState({
      id: (state) => { return state.route.params.id; },
    }),
    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },
    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },
    supportedCoins() {
      return Coin.all();
    },
    coinDenomination() {
      return this.supportedCoins.find((coin) => {
        return coin.name === this.wallet.name;
      }).denomination;
    },
    minConfirmations() {
      return this.supportedCoins.find((coin) => {
        return coin.name === this.wallet.name;
      }).minConfirmations;
    },
    coinSymbol() {
      return this.supportedCoins.find((coin) => { return coin.name === this.wallet.name; }).symbol;
    },
    latestPrice() {
      const prices = this.$store.getters['entities/latestPrice/find'](`${this.coinSymbol}_${this.selectedCurrency.code}`);
      if (prices) {
        return prices.data.PRICE;
      }
      return null;
    },

    /**
     * Prepends "To: " or "From: " into transaction title
     * @return {String}
     */
    paymentDirection() {
      if (this.data.sent) {
        let { receiver } = this.data;
        if (Array.isArray(receiver)) { [receiver] = receiver; }
        return `${this.$t('to')}: ${receiver}`;
      }

      let { sender } = this.data;
      if (Array.isArray(sender)) { [sender] = sender; }
      return `${this.$t('from')}: ${sender}`;
    },

    to() {
      let { receiver } = this.data;
      if (Array.isArray(receiver)) { [receiver] = receiver; }
      return receiver;
    },

    /**
     * Returns formatted date of the transaction
     * @return {String}
     */
    date() {
      const msToS = 1000;
      if (this.data.receivedTime) {
        return dateTranslater(new Date(this.data.receivedTime * msToS).valueOf(), 'DD MMMM HH:mm YYYY', this);
      }
      return dateTranslater(new Date(this.data.confirmedTime * msToS).valueOf(), 'DD MMMM HH:mm YYYY', this);
    },

    /**
     * Converts amount value (Number) into a String
     * @return {String}
     */
    amount() {
      const { value } = this.data;
      let inCoin = value;
      if (this.data.sent) { inCoin = -Math.abs(inCoin); }

      const amountInCoin = new AmountFormatter({
        amount: inCoin,
        rate: this.latestPrice,
        format: this.coinDenomination,
        prependPlusOrMinus: true,
        removeTrailingZeros: true,
      });

      if (this.latestPrice) {
        const amountInCurrency = new AmountFormatter({
          amount: inCoin,
          rate: this.latestPrice,
          format: '0,0[.]00',
          coin: this.wallet.name,
          prependPlusOrMinus: false,
          currency: this.selectedCurrency,
          toCurrency: true,
          toCoin: false,
          withCurrencySymbol: true,
        });

        return `${amountInCoin.getFormatted()} ${this.coinSymbol} (${amountInCurrency.getFormatted()})`;
      }
      return `${amountInCoin.getFormatted()} ${this.coinSymbol}`;
    },

    /**
     * Displays the fee
     * @return {String}
     */
    feeFormated() {
      const { fee } = this.data;
      const inCoin = fee;
      if (this.latestPrice) {
        let feeInCoin = new AmountFormatter({
          amount: inCoin,
          rate: this.latestPrice,
          format: this.coinDenomination,
          prependPlusOrMinus: false,
          removeTrailingZeros: true,
        });

        let amountInCurrency = new AmountFormatter({
          amount: feeInCoin.getFormatted(),
          rate: this.latestPrice,
          format: '0,0[.]00',
          coin: this.wallet.name,
          prependPlusOrMinus: false,
          currency: this.selectedCurrency,
          toCurrency: true,
          toCoin: false,
          withCurrencySymbol: true,
        });

        if (this.wallet.sdk === 'ERC20') {
          const parent = this.supportedCoins.find((coin) => {
            return coin.name === this.wallet.parentName;
          });
          const price = this.$store.getters['entities/latestPrice/find'](`${parent.symbol}_${this.selectedCurrency.code}`).data.PRICE;

          feeInCoin = new AmountFormatter({
            amount: inCoin,
            rate: price,
            format: parent.denomination,
            prependPlusOrMinus: false,
            removeTrailingZeros: true,
          });

          amountInCurrency = new AmountFormatter({
            amount: feeInCoin.getFormatted(),
            rate: price,
            format: '0,0[.]00',
            coin: this.wallet.parentName,
            prependPlusOrMinus: false,
            currency: this.selectedCurrency,
            toCurrency: true,
            toCoin: false,
            withCurrencySymbol: true,
          });
          return `${fee} ${parent.symbol} (${amountInCurrency.getFormatted()})`;
        }

        return `${fee} ${this.coinSymbol} (${amountInCurrency.getFormatted()})`;
      }

      return `${fee} ${this.coinSymbol}`;
    },

    /**
     * Returns the transaction status in a human friendly format
     * @return {String}
     */
    status() {
      if (this.data.confirmations > this.minConfirmations) { return this.$t('confirmed'); }
      if (this.data.confirmations > 0) { return this.$t('unconfirmed'); }
      return this.$t('pending');
    },
  },
};
</script>

<style>
.positive-amount .q-item__label,
.status.confirmed-tx {
  color: green;
}

.negative-amount .q-item__label,
.status.unconfirmed-tx {
  color: red;
}

.q-timeline-entry {
  padding-left: 1.5rem;
}

.q-item {
  padding: 0;
}

.q-timeline-content {
  padding-bottom: 0.5rem;
}

.q-timeline-title {
  margin-bottom: 0rem;
}

.q-timeline-dot:after {
  opacity: 0.1;
  left: 7px;
}

.q-timeline-dot:before, .q-timeline-dot:after {
    content: '';
    background: #4e677d;
    display: block;
    position: absolute;
}

.q-timeline-dot:before {
    border: 3px solid transparent;
    border-radius: 100%;
    height: 12px;
    width: 12px;
    top: 5px;
    left: 2px;
    transition: background 0.3s ease-in-out, border 0.3s ease-in-out;
}

.q-timeline {
  padding: 0.5rem;
}

.single-transaction-content {
  font-size: 0.8rem;
}

.q-timeline--dense--right
.q-timeline__entry {
  padding-left: 1.5rem;
}

.q-timeline__title {
  font-size: 1rem;
  padding-top: 0.1rem;
  margin-bottom: 0;
}

.q-expansion-item .q-item__label {
  font-size: 0.8rem;
}

.q-expansion-item .q-item {
  min-height: 1rem;
}

.q-expansion-item .q-card__section {
  padding: 0;
}

.q-expansion-item .q-card__section p {
  margin: 0;
  padding: 0.5rem;
  border-top: 1px solid whitesmoke;
}

.q-expansion-item__container {
  border-top: 1px solid white;
}

.break {
  word-break: break-all;
}
</style>
