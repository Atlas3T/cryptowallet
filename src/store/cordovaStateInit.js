/*
 * this function will consume the complete Vuex store object and
 * add event listeners to Cordova's deviceready
 */
export default function (store, Vue) {
  /*
   * getDeviceContacts will attempt to get all of the devices contacts
   * as well as all of the fields available on each contact
   * if it succedes its going to mutate the Vuex store
   * @TODO Uniform error message
   */
  function getDeviceContacts() {
    try {
      const options = new ContactFindOptions();
      options.multiple = true;
      const fields = ['*'];
      navigator.contacts.find(fields, (contacts) => {
        store.state.wallet.contacts = contacts;
      }, new Error('Failed to retrieve contacts from the device'), options);
    } catch (err) {
      Vue.config.errorHandler(err);
    }
  }

  document.addEventListener('deviceready', () => {
    getDeviceContacts();
  }, false);
}
