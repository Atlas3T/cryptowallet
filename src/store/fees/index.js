import { Model } from '@vuex-orm/core';

export default class fees extends Model {
  static entity = 'fees';

  static primaryKey = ['code'];

  static fields() {
    return {
      code: this.attr(''),
      timestamp: this.attr(''),
      data: this.attr(''),
    };
  }
}
