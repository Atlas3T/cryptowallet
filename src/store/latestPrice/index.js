import { Model } from '@vuex-orm/core';


export default class latestPrice extends Model {
    static entity = 'latestPrice';

    static primaryKey = ['coin', 'currency']


    static fields() {
      return {
        coin: this.attr(''),
        currency: this.attr(''),
        updated: this.attr(''),
        data: this.attr(''),
      };
    }
}
