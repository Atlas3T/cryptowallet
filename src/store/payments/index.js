/* eslint-disable */


const state = {
  balance: 0,
  transactions: [],
 /* transactions: [
    {
      ts: 1539522543102,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1535527543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 1,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539144763020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1529866783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1339527543150,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539527543166,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539557543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539134763020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1531141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 15439866783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1522144743020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539181123400,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1530066783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539517543120,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1522141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1529866553020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1336627543150,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539527548866,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539957543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 15439866783330,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 2222144743020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1566681123400,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
  ],*/
};

export default {
  namespaced: true,
  state,
};
