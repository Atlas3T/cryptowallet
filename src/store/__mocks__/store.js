import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
