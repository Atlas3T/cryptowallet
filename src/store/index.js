import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';

// import modules.
import account from './wallet/account';
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';

import cordovaStateInit from './cordovaStateInit';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, account);
database.register(Address, {});
database.register(Tx, {});
database.register(Wallet, {});
database.register(Coin, {});

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
});

/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
  },
  plugins: [vuexLocal.plugin, VuexORM.install(database)],
});

// load cordova data
cordovaStateInit(store, Vue);

export default store;
