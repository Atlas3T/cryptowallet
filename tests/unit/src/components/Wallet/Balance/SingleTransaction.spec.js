import { shallowMount } from '@vue/test-utils';
import SingleTransaction from '@/components/Wallet/Balance/SingleTransaction.vue';
import { localVue, i18n, createRouter } from '@/../tests/unit/helpers/setupLocalVue.js';
import { __createMocks as createStoreMocks } from '@/../tests/store/__mocks__/store.js';

describe('SingleTransaction.vue', () => {
  let storeMocks;
  let wrapper;
  let store;
  let router;

  const propsData = {
    data: {
      ts: 1539522543102,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    }
  };

  function wrapperInit (options) {
    return shallowMount(SingleTransaction, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: `/` });
    wrapper = wrapperInit({ i18n, router, localVue, store: storeMocks.store, propsData });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the qtimelineentry-stub', () => {
    expect(wrapper.contains('qtimelineentry-stub')).toBe(true);
  });

  it('renders the qcollapsible-stub within qtimelineentry-stub', () => {
    expect(wrapper.contains('qtimelineentry-stub qcollapsible-stub')).toBe(true);
  });

  it('computes and renders title correctly', () => {
    expect(wrapper.vm.$props.data.to).toBeTruthy();
    let title = `To: ${propsData.data.to}`;
    expect(wrapper.contains(`qtimelineentry-stub[title="${title}"]`)).toBe(true);

    const dataCopy = Object.assign({}, propsData.data);
    dataCopy.to = null;
    dataCopy.from = "Gal Anonim";
    wrapper.setProps({data: dataCopy});
    expect(wrapper.vm.$props.data.from).toBeTruthy();
    expect(wrapper.vm.$props.data.to).toBeFalsy();
    title = `From: ${wrapper.vm.$props.data.from}`;
    expect(wrapper.contains(`qtimelineentry-stub[title="${title}"]`)).toBe(true);
  });

  it('computes and renders date correctly', () => {
      const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
      const day = new Date(wrapper.vm.$props.data.ts).getDate();
      const month = monthNames[new Date(wrapper.vm.$props.data.ts).getMonth()];
      const date = `${day} ${month}`;
      expect(wrapper.contains(`qcollapsible-stub[label="${date}"]`)).toBe(true);
  });

  it('computes and renders the amount correctly', () => {
    let amount = wrapper.vm.$props.data.amount;
    expect(amount < 0).toBe(true);
    expect(wrapper.contains(`qcollapsible-stub.negative-amount[sublabel="&#45; ${Math.abs(amount)}"]`)).toBe(true);

    const dataCopy = Object.assign({}, propsData.data);
    dataCopy.amount = 6.22;
    wrapper.setProps({data: dataCopy});
    amount = wrapper.vm.$props.data.amount;
    expect(amount > 0).toBe(true);
    expect(wrapper.contains(`qcollapsible-stub.positive-amount[sublabel="&#43; ${amount}"]`)).toBe(true);
  });

  it('renders the received data point', () => {
    expect(wrapper.find('.received').text()).toBe(propsData.data.received);
  });

  it('renders the tx_hash data point', () => {
    expect(wrapper.find('.tx-hash').text()).toBe(propsData.data.tx_hash);
  });

  it('computes and renders the status correctly', () => {
    expect(wrapper.vm.$props.data.status).toBe(2);
    expect(wrapper.find('.status.confirmed-tx').text()).toBe('Confirmed');

    let dataCopy = Object.assign({}, propsData.data);
    dataCopy.status = 1;
    wrapper.setProps({data: dataCopy});
    expect(wrapper.vm.$props.data.status).toBe(1);
    expect(wrapper.find('.status.unconfirmed-tx').text()).toBe('Unconfirmed');

    dataCopy = Object.assign({}, propsData.data);
    dataCopy.status = 0;
    wrapper.setProps({data: dataCopy});
    expect(wrapper.vm.$props.data.status).toBe(0);
    expect(wrapper.find('.status').text()).toBe('Pending');
  });

  it('renders the tx_fee data point', () => {
    expect(wrapper.find('.tx-fee').text()).toBe(propsData.data.tx_fee);
  });
});
