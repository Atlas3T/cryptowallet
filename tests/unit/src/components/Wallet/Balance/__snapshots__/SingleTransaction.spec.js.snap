// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleTransaction.vue renders and matches snapshot 1`] = `
<qtimelineentry-stub
  side="left"
  tag="h3"
  title="To: Jack Black"
>
  <qcollapsible-stub
    class="negative-amount"
    label="14 October"
    sublabel="&#8722; 4.97"
  >
    <div>
      <p>
        
        Received:
        <br />
         
        <span
          class="received"
        >
          1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE
        </span>
      </p>
       
      <p>
        
        TX Hash:
        <br />
         
        <span
          class="tx-hash"
        >
          1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE
        </span>
      </p>
       
      <p>
        
        Status:
        <br />
         
        <span
          class="status confirmed-tx"
        >
          
          Confirmed
        
        </span>
      </p>
       
      <p>
        
        TX Fee:
        <br />
         
        <span
          class="tx-fee"
        >
          0.000000012
        </span>
      </p>
    </div>
  </qcollapsible-stub>
</qtimelineentry-stub>
`;
